File: src-tauri/src/agent_orchestration/types_test.rs
//! stages_from_bindings / apply_stage_bindings 单测

use crate::agent_orchestration::types::{
    apply_stage_bindings, default_stage_specs, stages_from_bindings, AgentStageBindingInput,
};
use crate::shared_event_log::canonical::types::CanonicalProviderProfileSource;
use crate::shared_session_v2::{EngineType, ExecutionTargetInput};

fn target(engine: EngineType) -> ExecutionTargetInput {
    ExecutionTargetInput {
        engine,
        provider_profile_id: None,
        model_catalog_entry_id: Some("m1".into()),
        model: Some("m1".into()),
        reasoning_effort: Some("medium".into()),
        provider_profile_name_snapshot: Some("local".into()),
        provider_profile_source: Some(CanonicalProviderProfileSource::Local),
        runtime_capability_fingerprint: None,
    }
}

#[test]
fn stages_from_bindings_preserves_n_stages_and_role_prompt() {
    let default = target(EngineType::Codex);
    let bindings = vec![
        AgentStageBindingInput {
            id: "plan".into(),
            target: target(EngineType::Claude),
            title: Some("规划".into()),
            role_prompt: Some("only plan".into()),
            access_mode: Some("read-only".into()),
            requires_approval: Some(true),
            persona_agent_id: Some("a1".into()),
            persona_agent_name: Some("小张".into()),
            persona_agent_icon: None,
        },
        AgentStageBindingInput {
            id: "implement".into(),
            target: target(EngineType::Codex),
            title: Some("实现".into()),
            role_prompt: None,
            access_mode: Some("current".into()),
            requires_approval: Some(false),
            persona_agent_id: None,
            persona_agent_name: None,
            persona_agent_icon: None,
        },
        AgentStageBindingInput {
            id: "test-harden".into(),
            target: target(EngineType::Claude),
            title: Some("测试加固".into()),
            role_prompt: Some("add tests".into()),
            access_mode: Some("current".into()),
            requires_approval: Some(false),
            persona_agent_id: None,
            persona_agent_name: None,
            persona_agent_icon: None,
        },
        AgentStageBindingInput {
            id: "review".into(),
            target: target(EngineType::Grok),
            title: Some("审查".into()),
            role_prompt: None,
            access_mode: Some("read-only".into()),
            requires_approval: Some(false),
            persona_agent_id: None,
            persona_agent_name: None,
            persona_agent_icon: None,
        },
    ];
    let stages = stages_from_bindings(&default, &bindings);
    assert_eq!(stages.len(), 4);
    assert_eq!(stages[0].id, "plan");
    assert_eq!(stages[0].role_prompt.as_deref(), Some("only plan"));
    assert_eq!(stages[0].persona_agent_name.as_deref(), Some("小张"));
    assert!(stages[0].requires_approval);
    assert_eq!(stages[2].id, "test-harden");
    assert_eq!(stages[2].title, "测试加固");
    assert!(!stages[2].requires_approval);
}

#[test]
fn apply_stage_bindings_rebuilds_when_rich_metadata_present() {
    let default = target(EngineType::Codex);
    let base = default_stage_specs(&default);
    assert_eq!(base.len(), 3);
    let bindings = vec![
        AgentStageBindingInput {
            id: "a".into(),
            target: target(EngineType::Claude),
            title: Some("A".into()),
            role_prompt: Some("ra".into()),
            access_mode: None,
            requires_approval: Some(false),
            persona_agent_id: None,
            persona_agent_name: None,
            persona_agent_icon: None,
        },
        AgentStageBindingInput {
            id: "b".into(),
            target: target(EngineType::Codex),
            title: Some("B".into()),
            role_prompt: None,
            access_mode: None,
            requires_approval: Some(false),
            persona_agent_id: None,
            persona_agent_name: None,
            persona_agent_icon: None,
        },
    ];
    let stages = apply_stage_bindings(base, &bindings);
    assert_eq!(stages.len(), 2);
    assert_eq!(stages[0].title, "A");
}

File: src/features/multi-agent/templates/types.test.ts
import { describe, expect, it } from "vitest";

import {
  composeStageRolePrompt,
  mergeTarget,
  normalizeAgentTargetSource,
  templateToStageBindings,
  type CollaborationTemplate,
} from "./types";
import type { AgentExecutionTarget } from "../types";

const sessionCodex: AgentExecutionTarget = {
  engine: "codex",
  providerProfileId: null,
  modelCatalogEntryId: "gpt-5.3-codex",
  model: "gpt-5.3-codex",
  reasoningEffort: "medium",
  providerProfileNameSnapshot: "本地配置",
  providerProfileSource: "disk",
  runtimeCapabilityFingerprint: null,
};

const completeClaude: AgentExecutionTarget = {
  engine: "claude",
  providerProfileId: null,
  modelCatalogEntryId: "claude-sonnet-4-6",
  model: "claude-sonnet-4-6",
  reasoningEffort: "high",
  providerProfileNameSnapshot: "本地配置",
  providerProfileSource: "disk",
  runtimeCapabilityFingerprint: null,
};

describe("mergeTarget", () => {
  it("keeps complete stage target", () => {
    const merged = mergeTarget(completeClaude, sessionCodex);
    expect(merged.engine).toBe("claude");
    expect(merged.model).toBe("claude-sonnet-4-6");
  });

  it("does not mix cross-engine incomplete target with session model", () => {
    const incompleteClaude = {
      ...completeClaude,
      model: null,
      modelCatalogEntryId: null,
      providerProfileNameSnapshot: null,
      providerProfileSource: null,
      engine: "claude" as const,
    };
    const merged = mergeTarget(incompleteClaude, sessionCodex);
    // 安全回退会话 target，避免 claude+codex 非法组合
    expect(merged.engine).toBe("codex");
    expect(merged.model).toBe("gpt-5.3-codex");
  });

  it("fills same-engine gaps from session", () => {
    const partial = {
      ...sessionCodex,
      model: null,
      modelCatalogEntryId: null,
      reasoningEffort: "high" as string | null,
    };
    const merged = mergeTarget(partial, sessionCodex);
    expect(merged.engine).toBe("codex");
    expect(merged.model).toBe("gpt-5.3-codex");
    expect(merged.reasoningEffort).toBe("high");
  });
});

describe("normalizeAgentTargetSource", () => {
  it("maps disk to local for backend validate", () => {
    const next = normalizeAgentTargetSource(sessionCodex);
    expect(next.providerProfileSource).toBe("local");
  });
});

describe("templateToStageBindings", () => {
  it("emits N stages without collapsing to 3", () => {
    const tpl: CollaborationTemplate = {
      id: "fix4",
      name: "修复流水线",
      description: "",
      builtin: true,
      version: 1,
      updatedAt: 0,
      stages: [
        {
          id: "plan",
          title: "规划",
          rolePrompt: "plan me",
          accessMode: "read-only",
          requiresApproval: true,
          target: { engine: "claude" } as AgentExecutionTarget,
        },
        {
          id: "implement",
          title: "实现",
          rolePrompt: "",
          accessMode: "current",
          requiresApproval: false,
          target: { engine: "codex" } as AgentExecutionTarget,
        },
        {
          id: "test-harden",
          title: "测试加固",
          rolePrompt: "tests",
          accessMode: "current",
          requiresApproval: false,
          target: { engine: "claude" } as AgentExecutionTarget,
        },
        {
          id: "review",
          title: "审查",
          rolePrompt: "",
          accessMode: "read-only",
          requiresApproval: false,
          target: { engine: "grok" } as AgentExecutionTarget,
        },
      ],
    };
    const bindings = templateToStageBindings(tpl, sessionCodex);
    expect(bindings).toHaveLength(4);
    expect(bindings.map((b) => b.id)).toEqual([
      "plan",
      "implement",
      "test-harden",
      "review",
    ]);
    expect(bindings[0]?.requiresApproval).toBe(true);
    expect(bindings[0]?.rolePrompt).toBe("plan me");
    // 首段默认 full；非首段默认 summary
    expect(bindings[0]?.upstreamFeedMode).toBe("full");
    expect(bindings[1]?.upstreamFeedMode).toBe("summary");
    expect(bindings[3]?.upstreamFeedMode).toBe("summary");
    // incomplete cross-engine → 回退会话 target 且 source 归一 local
    expect(bindings[0]?.target.engine).toBe("codex");
    expect(bindings[0]?.target.providerProfileSource).toBe("local");
  });

  it("sends persona prompt separately from rolePrompt (not merged into role)", () => {
    const tpl: CollaborationTemplate = {
      id: "with-agent",
      name: "带智能体",
      description: "",
      builtin: false,
      version: 1,
      updatedAt: 0,
      stages: [
        {
          id: "plan",
          title: "规划",
          rolePrompt: "只做计划",
          accessMode: "read-only",
          requiresApproval: true,
          personaAgentId: "agent-xz",
          personaAgentName: "小张",
          personaAgentIcon: "codicon-robot",
          personaAgentPrompt: "你是资深工程师，偏可验证交付。",
          target: completeClaude,
        },
      ],
    };
    const bindings = templateToStageBindings(tpl, sessionCodex);
    expect(bindings[0]?.rolePrompt).toBe("只做计划");
    expect(bindings[0]?.personaPrompt).toBe("你是资深工程师，偏可验证交付。");
    expect(bindings[0]?.personaAgentName).toBe("小张");
    // 禁止把正文拼进 rolePrompt（幕布/投影泄漏）
    expect(bindings[0]?.rolePrompt).not.toContain("你是资深工程师");
  });
});

describe("composeStageRolePrompt", () => {
  it("returns null when empty", () => {
    expect(
      composeStageRolePrompt({
        id: "a",
        title: "t",
        rolePrompt: "",
        accessMode: "current",
        requiresApproval: false,
        target: { engine: "claude" } as AgentExecutionTarget,
      }),
    ).toBeNull();
  });

  it("returns stage body only without agent name header", () => {
    expect(
      composeStageRolePrompt({
        id: "a",
        title: "t",
        rolePrompt: "本步约束",
        accessMode: "current",
        requiresApproval: false,
        personaAgentName: "小张",
        personaAgentPrompt: "人设正文",
        target: { engine: "claude" } as AgentExecutionTarget,
      }),
    ).toBe("本步约束");
  });
});
